### DISCUSSION TITLE
BatchInfo update

### DISCUSSION THREAD

**Discussion by JiriOgurek** [Mimetis/Dotmim.Sync#817] [Q&A]
Hi, you have suggested to remove deleted rows using this code:
```
agent.LocalOrchestrator.OnTableChangesSelected(async tcsa =>
{
    foreach (var bpi in tcsa.BatchPartInfos)
    {
        var table = await tcsa.BatchInfo.LoadBatchPartInfoAsync(bpi);

        foreach (var row in table.Rows.ToArray())
        {
            if (row.RowState == DataRowState.Deleted)
            {
                table.Rows.Remove(row);
            }
        }

        await tcsa.BatchInfo.SaveBatchPartInfoAsync(bpi, table);
    }
});
```
However, it's not working anymore since BatchInfo class has changed, would you be so kind and suggest new approach, please?

**Comment by Mimetis** [comment]
Hi @JiriOgurek 

Thanks for the feedback

I have a small issue regarding this code, that will be fixed in next release.

That being said, I think there is a more comfortable way to remote rows you don't want to sync.
First of all, you can specify a SyncDirection to your table, from your setup

``` cs
var setup = new SyncSetup("ProductCategory", "Product");
setup.Tables["ProductCategory"].SyncDirection = SyncDirection.DownloadOnly;
```

If it's not fitting your scenario, you can filter the rows, during the read of your database, like that:

``` cs
agent.RemoteOrchestrator.OnRowsChangesSelected(args =>
{
    if (args.SchemaTable.TableName == "ProductCategory" && args.SyncRow.RowState == SyncRowState.Deleted)
        args.SyncRow = null;
});

```

If you don't want to filter rows during read time, but more the rows are serialized locally (and connection closed), you can use your code, rewrite like this:

> **Note**: This code will work in next release since I have a small issue here

``` cs
agent.RemoteOrchestrator.OnTableChangesSelected(async args =>
{
    foreach (var bpi in args.BatchPartInfos)
    {
        var table = await agent.RemoteOrchestrator.LoadTableFromBatchPartInfoAsync(args.BatchInfo, bpi);

        foreach (var row in table.Rows.ToArray())
            if (row.RowState == SyncRowState.Deleted)
                table.Rows.Remove(row);

        await agent.RemoteOrchestrator.SaveTableToBatchPartInfoAsync(args.BatchInfo, bpi, table);
    }
});

```

Let me know if it's working for you

**Reply by JiriOgurek** [reply]
Hi @Mimetis,

thank you for your kind help. Yes, I do specify UploadOnly, and I've tried 2nd block of code for LocalOrchestator, since 3rd code is compilable but not working right now (as you mentioned). When I try to run it I get exception text: "This version needs a client manual upgrade." What shall I do, please?

**Reply by Mimetis** [reply]
Seems you are using the last version of DMS (0.9.7) but you have an older version of the client / server databases, right ?
I mean, you have created the sync with an older version, and the scopes tables created by DMS have changed since yet.

You need to upgrade. https://github.com/Mimetis/Dotmim.Sync/discussions/802#discussioncomment-3594681
If you are just testing, I think you should start from scratch

**Comment by JiriOgurek** [comment]
Hi @Mimetis, 

I've delayed this update to the end of the year (when we will setup new database) to avoid the upgrade troubles (to start from scratch). However upon first attempt to sync I get this error message: "No Schema in your server scopeInfo DefaultScope", what's wrong, please? Take a look at the source code diff when upgrading v0.9.4 --> 0.9.7, please.

![diff](https://user-images.githubusercontent.com/49657334/205058540-c2139de5-847b-4e87-b8bb-2ca4ab86d59a.png)

**Reply by Mimetis** [reply]
Do you try an upgrade or are you testing from scratch from a server database without anything related to `DMS` ?

Looking at your code, do not try a `RemoteOrchestrator.UpgradeAsync` if you are testing a new version (just in case)
Dont use `SynchronousProgress` if you are in a UI program (like WPF or Windows Forms) Prefer a simple `Progress<T>` object

The message "No Schema in your server scopeInfo DefaultScope" means you do not have a schema to sync from your server database.
Is your server database empty ?

You have to pass your setup instance in the `SynchronizeAsync()` method, if you never did before
(you need to do it at least one time. )

**Comment by JiriOgurek** [comment]
Hi @Mimetis, 

thank you for your kind help. I've commented out the upgrade attempt and passed setup to the SynchronizeAsync() method. The first sync init went fine, trackings are created, however second sync run (with new data added) throws exception: "Object reference not set to an instance of an object." Do you have a clue what's wrong, please?

**Reply by Mimetis** [reply]
Hey @JiriOgurek 
Make an upgrade to` 0.9.8 beta` version (a minor version that adds a lot of logs and errors messages. If you are already on` 0.9.7`, there is nothing to do to upgrade to `0.9.8 beta`) to have a full stack trace of the error raised.
Post the error stack trace here to help me figure it out

**Comment by JiriOgurek** [comment]
Hi @Mimetis,

I have upgraded to beta as you suggested.

First sync init OK, tracking tables created. One batch of data added to local and sync run seems to be fine (see uploaded records count in the screenshot), however in fact there are no new records at the remote server. When I add more data batches and try to sync new exception raises: "Cannot write to a closed TextWriter." Any suggestions, please?

![uploaded](https://user-images.githubusercontent.com/49657334/205307563-41eba028-410e-48a4-a9e1-dc4170c0d6e7.png)
![exception](https://user-images.githubusercontent.com/49657334/205307548-20858a27-1492-4739-97b9-9060e9400a46.png)

**Reply by Mimetis** [reply]
Can you share the code ?

**Comment by JiriOgurek** [comment]
@Mimetis, sure. on the left side it's working v0.9.4; on the right it's latest beta version throwing exceptions.
![screenshot](https://user-images.githubusercontent.com/49657334/205570753-00f262e8-8220-435c-9b63-83283ff2ba93.png)

**Reply by Mimetis** [reply]
I'm sorry, i'm not able to reproduce the error.
Can you prepare a Github project sample where you are reproducing the error ? 
Will be easier for me to debug

**Comment by JiriOgurek** [comment]
@Mimetis, so source code seems to be fine, please?

**Reply by Mimetis** [reply]
Dont use `SynchronousProgress` if you are not using a console application (like WPF or Windows Forms) Prefer a simple `Progress<T>` object
Everything else seems to be fine, from what I see.

Do you have the same regional settings from the server and the client ?

**Comment by JiriOgurek** [comment]
@Mimetis, you mean Windows regional settings? Probably not, is there a solution?

**Comment by JiriOgurek** [comment]
@Mimetis, I've tried comment out the LocalOrchestrator.OnTableChangesSelected event method and it works as expected :-) The only problem is, I have to remove local delete trackings since I have to preserve all server records, can you suggest proper way how to acomplish this, please?

**Reply by Mimetis** [reply]
Are using `SyncDirection.DownloadOnly;` ? 

``` cs
setup.Tables["ProductCategory"].SyncDirection = SyncDirection.DownloadOnly;
```
This way, you are not uploading **anything** to the server.

If you need to upload `Insert` / `Updates` and no `Deletes`, you can do something like, on the client side:

``` cs
// Bidirectional is the default value. Just mention it here to see the difference with last solution
setup.Tables["ProductCategory"].SyncDirection = SyncDirection.Bidirectional; 

// Intercepting rows during get changes, before serializing on disk
agent.LocalOrchestrator.OnRowsChangesSelected(args =>
{
    // if we have a deleted row, we set the row content to null to not serialize it.
    if (args.SyncRow.RowState == SyncRowState.Deleted)
        args.SyncRow = null;
});
```

let me know if this solution works for you

**Comment by JiriOgurek** [comment]
@Mimetis, I am using UploadOnly, I don't want do download anything from remote server, plus I don't want to upload local deletes. I've been using this code successfully till v0.9.4. However BatchInfo structure has been changed. Can you post how to remove local deletes in UploadOnly scenario, please?
```
agent.LocalOrchestrator.OnTableChangesSelected(async tcsa =>
{
    foreach (var bpi in tcsa.BatchPartInfos)
    {
        var table = await tcsa.BatchInfo.LoadBatchPartInfoAsync(bpi);

        foreach (var row in table.Rows.ToArray())
        {
            if (row.RowState == DataRowState.Deleted)
            {
                table.Rows.Remove(row);
            }
        }

        await tcsa.BatchInfo.SaveBatchPartInfoAsync(bpi, table);
    }
});
```

**Reply by Mimetis** [reply]
Did you see my last comment ?
I thought it was pretty clear ...

``` cs
// Intercepting rows during get changes, before serializing on disk
agent.LocalOrchestrator.OnRowsChangesSelected(args =>
{
    // if we have a deleted row, we set the row content to null to not serialize it.
    if (args.SyncRow.RowState == SyncRowState.Deleted)
        args.SyncRow = null;
});
```

What are you not understanding here ?

**Comment by JiriOgurek** [comment]
@Mimetis, I am sorry, I thought it's related to Download scenario only, I do apologize. It seems to be working fine :-) Thank you for your kind help, I will run more tests now.

**Comment by JiriOgurek** [comment]
@Mimetis, with small updates everything works as expected, when I add thousands of records (in one batch) I get this exception "Cannot write to a closed TextWriter", can you help me, please?

![exception](https://user-images.githubusercontent.com/49657334/205904507-aed8dba3-2bff-468a-bc5e-d04182d32474.png)

**Reply by Mimetis** [reply]
I cant until you are able to share a sample where you are reproducing the error.

**Reply by JiriOgurek** [reply]
@Mimetis, I am sorry, there's no easy way how to create this project sample, anyway I have debugged the exception, can you take a look, please? Just my guess - trying to close already closed TextWriter?

![scr1](https://user-images.githubusercontent.com/49657334/205915116-e7f740c6-9c7d-4132-9d6f-6d0ff11664fc.png)
![scr2](https://user-images.githubusercontent.com/49657334/205915120-66bdb5dd-2998-483e-b3bb-e54a54716990.png)
![scr3](https://user-images.githubusercontent.com/49657334/205915122-f5e49c5f-7d80-42f4-8340-56ae32e9ff69.png)

**Reply by Mimetis** [reply]
Damn... that's weird !!!

I've just committed a patch on the master branch.
I tried to lock the writer, before trying to close it, and I fallback to synchronous call.

Can you download the last source code (git clone or direct download), and try it directly in your code and see if the problem still occurs?

**Comment by JiriOgurek** [comment]
@Mimetis, sure! I will do so and let you know.

**Comment by JiriOgurek** [comment]
@Mimetis, congratulations! :-) It's working as expected, I will perform additional tests now.
